src/analytics/internal/sendToServer/SendToServer.js

```diff
@@ -0,0 +1,130 @@
+import { idleManager } from 'src/util/idle'
+import { SHOULD_TRACK_STORAGE_KEY as SHOULD_TRACK } from 'src/options/privacy/constants'
+import { USER_ID, generateTokenIfNot } from 'src/util/generate-token'
+import { installTimeStorageKey } from 'src/imports/background'
+
+class SendToServer {
```

Comment:
This class has a lot of methods in-common with the class in `src/analytics/analytics` module. As they're not TS, I don't think there's much of a point in making an abstract class they both extend right now, but the main issue I have is that the properties and methods that are named the same have different behaviours or types.

While the existing class is far from perfect, it's really important to be consistent here if you re-use the same names and overall class layout.

Some examples:
- `_pool` prop is an `Array` in here, while in the existing Analytics class it's a `Set<URLSearchParams>`
- the same XHR header values are rewritten multiple times in this class while in the existing class they are on the `static JSON_HEADER` prop
- `_formReqParams` method here creates a custom object for sending in your XHRs, while in the existing class it constructs a [`URLSearchParams`](https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams) object for storing in the pool (it also contains a fairly ambiguous `params` arg which switches behaviour between sending a single req and a bulk req)

Reply:
Here the things are different, so that's why I took the same but the functionality is different.
Here we have to send the data as a json in which it will contain the deep object, that's why I am not able to use `URLSearchParams`, therefore I make the whole object serializable and send it to as a json with stringify. 
Here `_pool` is the array becuase we have to make the array of all the events and will serializable at the end. :slightly_frowning_face:

Reply:
Yeah sorry, `URLSearchParams` wouldn't be appropriate here - I didn't think about it's purpose in the old class. But in fact that means there's no real point to have `_formReqParams` method then; the purpose of this in the existing Analytics class is to simply to construct the request shape, for a **single** event, for storing in the `_pool`. 

In your new class the `_formReqParams` method is doing a completely different thing: checking the user ID, requesting a new one if none exists locally, and then batching up **all** events already stored in the `_pool`, attaching the user ID, ready for sending in the req data. All the ID related logic in there can probably be moved to your `userId` method. The batching of pool data is done in the `_serializePoolReqs` in the existing Analytics class (that should really have the `JSON.stringify` call in there too, as that's part of the serialization). You can extract that logic out of the `_formReqParams` into a new `_serializePoolReqs` method, which you may also want to call `userId` method from to get the ID (or accept it as a param), then you can remove `_formReqParams`.

`_pool` should still be a `Set` here; there's no reason for the inconsistency. You can just destruct it into a new array like `[...this._pool]` when you need to use the stored data (see `_serializePoolReqs` in the existing Analytics class).

You will also need to slightly update `_sendReq`. It just needs to get the ID, then serialize it with the `event` it already has. Maybe you could do this with the `_serializePoolReqs` method, but you would need some conditional logic. TBH I'm not sure if the `_sendReq` method is that important; in the existing class there is only one use-case for it: when user "disables tracking" the requirement was to send off one final event so we know they disabled tracking :/ 

One final thing I just saw is you're checking `res.success`, but I'm pretty sure it should be `res.ok`. Docs: https://developer.mozilla.org/en-US/docs/Web/API/Response/ok